modules/core/shared/src/main/scala/retry/package.scala

```diff
@@ -28,32 +28,37 @@ package object retry {
   def noop[M[_]: Monad, A]: (A, RetryDetails) => M[Unit] =
     (_, _) => Monad[M].pure(())
 
+  /**
+    * A handler that inspects the result of an action and decides
+    * what to do next.
+    * This is also a good place to do any logging.
+    */
+  type ResultHandler[F[_], Res, A] = (Res, RetryDetails) => F[HandlerDecision[F, A]]
```

Comment:
We could split this into 3 similar types, providing a bit more clarity in exchange for redundancy:

```scala
type ResultHandler[F[_], A]           = (A, RetryDetails) => F[HandlerDecision[F, A]]
type ErrorHandler[F[_], A, E]         = (E, RetryDetails) => F[HandlerDecision[F, A]]
type ResultOrErrorHandler[F[_], A, E] = (Either[E, A], RetryDetails) => F[HandlerDecision[F, A]]
```